shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/EpochBoundary.hs

```diff
@@ -79,7 +80,19 @@ getStakeHK _ = Nothing
 
 aggregateOuts :: Crypto crypto => UTxO crypto -> Map (Addr crypto) Coin
 aggregateOuts (UTxO u) =
-  Map.fromListWith (+) (map (\(_, TxOut a c) -> (a, c)) $ Map.toList u)
+  -- Map.fromListWith (+) (map (\(_, TxOut a c) -> (a, c)) $ Map.toList u)
+  Map.foldr accum Map.empty u
+    where accum (TxOut a c) ans = Map.insertWith (+) a c ans
+
+aggregateStakes::  Crypto crypto => Map Ptr (Credential 'Staking crypto) -> UTxO crypto -> [(Credential 'Staking crypto, Coin)]
+aggregateStakes ptrs (UTxO u) =
+  Map.foldr accum [] u
+    where accum (TxOut (Addr _ _ (StakeRefBase hk)) c) ans = (hk,c):ans
+          accum (TxOut (Addr _ _ (StakeRefPtr p)) c) ans =
+             case Map.lookup p ptrs of
+               Just cred -> (cred,c):ans
+               Nothing -> ans
+          accum _other ans = ans
```

Comment:
So here we skip the `AddrBootstrap` case:
```haskell
data Addr crypto
  = Addr !Network !(PaymentCredential crypto) !(StakeReference crypto)
  | AddrBootstrap !(BootstrapAddress crypto)
```
All translated Byron addresses are `AddrBootstrap`'s, so (at least in the beginning) the majority of all `Addr`s will be `AddrBootstrap`s. In other words, we're wasting most of our time deserialising (see my other comment about `deserialiseAddr`) Byron addresses that we'll ignore anyway.

Idea: in `accum`:
1. Pattern match on the `TxOutCompact` constructor to avoid calling `deserialiseAddr` unconditionally for each address (`ShortByteString`).
2. Write a function `deserialiseStakeReference` (or a better name) with as type: `ByteString -> Maybe (StakeReference crypto)` which returns `Nothing` for `AddrBootstrap` (just test a bit to see whether it's Byron and then skip it) and only deserialises the `StakeReference` in case it's a Shelley `Addr` (you could deserialise the `Network` and `PaymentCredential` as well if needed elsewhere). Maybe throw in an `Either` there, because we're conflating "deserialisation failure" with "not a Shelley Addr".
3. Call pattern match on the result of calling `deserialiseStakeReference` on the `ShortByteString` argument to `TxOutCompact` to get the same three cases you currently have in `accum`.

Reply:
So all the above comments have been explored. I wrote a dozen different algorithms to try and make aggregateOuts (and its successors) faster. All of them were benchmarked and the fasted were wriiten into the latest version. We ran all the tests and all the benchmarks, every thig is working.